src/Compiler/Driver/CompilerDiagnostics.fs

```diff
@@ -381,943 +376,1082 @@ let IsWarningOrInfoEnabled (diagnostic, severity) n level specificWarnOn =
+
+        | ConstraintSolverTypesNotInSubsumptionRelation (denv, ty1, ty2, m, m2) ->
+            // REVIEW: consider if we need to show _cxs (the type parameter constraints)
+            let ty1, ty2, cxs = NicePrint.minimalStringsOfTwoTypes denv ty1 ty2
+            os.AppendString(ConstraintSolverTypesNotInSubsumptionRelationE().Format ty2 ty1 cxs)
+
+            if m.StartLine <> m2.StartLine then
+                os.AppendString(SeeAlsoE().Format(stringOfRange m2))
+
+        | ConstraintSolverError (msg, m, m2) ->
+            os.AppendString msg
+
+            if m.StartLine <> m2.StartLine then
+                os.AppendString(SeeAlsoE().Format(stringOfRange m2))
+
+        | ErrorFromAddingTypeEquation (g, denv, ty1, ty2, ConstraintSolverTypesNotInEqualityRelation (_, ty1b, ty2b, m, _, contextInfo), _) when
+            typeEquiv g ty1 ty1b && typeEquiv g ty2 ty2b
+            ->
+            let ty1, ty2, tpcs = NicePrint.minimalStringsOfTwoTypes denv ty1 ty2
+
+            match contextInfo with
+            | ContextInfo.IfExpression range when equals range m -> os.AppendString(FSComp.SR.ifExpression (ty1, ty2))
+
+            | ContextInfo.CollectionElement (isArray, range) when equals range m ->
+                if isArray then
+                    os.AppendString(FSComp.SR.arrayElementHasWrongType (ty1, ty2))
+                else
+                    os.AppendString(FSComp.SR.listElementHasWrongType (ty1, ty2))
+
+            | ContextInfo.OmittedElseBranch range when equals range m -> os.AppendString(FSComp.SR.missingElseBranch (ty2))
+
+            | ContextInfo.ElseBranchResult range when equals range m -> os.AppendString(FSComp.SR.elseBranchHasWrongType (ty1, ty2))
+
+            | ContextInfo.FollowingPatternMatchClause range when equals range m ->
+                os.AppendString(FSComp.SR.followingPatternMatchClauseHasWrongType (ty1, ty2))
+
+            | ContextInfo.PatternMatchGuard range when equals range m -> os.AppendString(FSComp.SR.patternMatchGuardIsNotBool (ty2))
+
+            | ContextInfo.TupleInRecordFields ->
+                os.AppendString(ErrorFromAddingTypeEquation1E().Format ty2 ty1 tpcs)
+                os.AppendString(Environment.NewLine + FSComp.SR.commaInsteadOfSemicolonInRecord ())
+
+            | _ when ty2 = "bool" && ty1.EndsWithOrdinal(" ref") ->
+                os.AppendString(ErrorFromAddingTypeEquation1E().Format ty2 ty1 tpcs)
+                os.AppendString(Environment.NewLine + FSComp.SR.derefInsteadOfNot ())
+
+            | _ -> os.AppendString(ErrorFromAddingTypeEquation1E().Format ty2 ty1 tpcs)
+
+        | ErrorFromAddingTypeEquation (_, _, _, _, (ConstraintSolverTypesNotInEqualityRelation (_, _, _, _, _, contextInfo) as e), _) when
+            (match contextInfo with
+             | ContextInfo.NoContext -> false
+             | _ -> true)
+            ->
+            OutputExceptionR os e
+
+        | ErrorFromAddingTypeEquation (_,
```

Comment:
In future, if we use discards, shall we consider putting them on one line?
For example:
```fs
        | ErrorFromAddingTypeEquation (_,_,_,_,
		                                             (ConstraintSolverTypesNotInSubsumptionRelation _
		                                             | ConstraintSolverError _ as e),
		                                             _) -> OutputExceptionR os e
```

Reply:
Possibly yes.  May notice that a lot in CheckExpressions and CheckDeclarations.  

In this case it should probably just be written as two clauses

```fsharp
        | ErrorFromAddingTypeEquation (error=(ConstraintSolverTypesNotInSubsumptionRelation _ as e))
        | ErrorFromAddingTypeEquation (error=(ConstraintSolverError _ as e)) ->
```